File: Programmi in C/Setteemezzo.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int matta (int, int);
float lancio ();

int main(){
	srand(time(NULL));
	
	//dichiarazione e assegnazione variabili
	float carta, cartaMazziere, puntoGioc, puntoMazz;
	char risp;
	int flagMattaMazziere = 0, flagMatta = 0;
	
	puts("REGOLE SETTE E MEZZO:");
	printf("Valore carte:\nDa 1 a 7 hanno il loro valore;\n8, 9 e 10 valgono 0.5.\nPer indicare la matta viene usato l'11.\n");
	printf("Ricevi una carta dal mazziere, se questa e' un 4 la puoi cambiare.\n");
	printf("Se ricevi l'11 (cioe' la matta) alla prima mano o dopo che hai cambiato il 4,\npuoi scegliere un valore da 0.5 a 7 con cui giocarla.");
	printf("Potrai scegliere se ricevere altre carte per avvicinarti a 7.5,\nse vai oltre questo punteggio 'sballi' quindi perdi la partita.");
	printf("Se alla fine della mano tu e il mazziere avrete carte uguali vince il mazziere, tranne se tu hai la matta.\n");
	puts("-------------------------------------------------------------------------------");
	carta = lancio();
	printf("LA TUA CARTA -> %.1f\n", carta);
	//gestione del 4 come prima carta
	if(carta == 4){
		puts("Vuoi cambiare carta?");
		scanf(" %c", &risp);
		if(risp == 's'){
			carta = lancio();
		}
		printf("LA TUA CARTA INZIALE -> %.1f\n", carta);
	}
	//gestione della matta come prima carta
	if(carta == 11){
		carta = matta(carta, flagMatta);
		printf("LA TUA CARTA INZIALE -> %.1f\n", carta);
	}
	//stampa carta con cui si inizia a giocare
	puntoGioc = carta;
	
	//Giocata dell'utente
	printf("Vuoi un'altra carta (s/n) ?");
	scanf(" %c", &risp);
	if(risp == 's'){	
		while(risp == 's'){
			if(puntoGioc < 7.5){
				carta = lancio();
				printf("\tCarta ricevuta -> %.1f\n", carta);
				if(carta == 11){
					carta = matta(carta, flagMatta);
				}
				puntoGioc += carta;	
				printf("\tPunteggio corrente -> %.1f\n", puntoGioc);
				printf("\tVuoi un'altra carta (s/n) ?");
				scanf("\t %c", &risp);
			}else{
				puts("\tNon puoi ricevere altre carte.");
				risp = 'n';
			}
		}	
	}
	
	printf("IL TUO PUNTEGGIO -> %.1f\n", puntoGioc);
	if(puntoGioc > 7.5){
		puts("Hai sballato!\nParita vinta dal mazziere.");
	}else{
		//Giocata del Mazziere
		cartaMazziere = lancio();
		printf("\nCARTA INIZIALE DEL MAZZIERE -> %.1f\n", cartaMazziere);
		if(cartaMazziere < puntoGioc){
			puntoMazz = cartaMazziere;
			do{
				cartaMazziere = lancio();
				if(cartaMazziere == 11){
					flagMattaMazziere = 1;
					if(puntoMazz == 7){
						cartaMazziere = 0.5;
					}else{
						cartaMazziere = (int) (7 - puntoMazz);
					}
				}
				puntoMazz += cartaMazziere;
			}while(puntoMazz < puntoGioc);
			printf("PUNTEGGIO MAZZIERE -> %.1f\n", puntoMazz);
			
			if(puntoMazz <= 7.5){
				if(puntoMazz == puntoGioc){
					if(flagMattaMazziere == 0){
						if(flagMatta == 1){
							printf("Hai vinto, perche' hai la matta, con il punteggio di %.1f", puntoGioc);
						}else{
							puts("Hai perso!");
						}
					}else{
						puts("Hai perso, perche' il mazziere ha la matta.");
					}
				}else{
					printf("Hai perso!\nPartita vinta dal mazziere con il punteggio di %.1f.", puntoMazz);
				}
			}else{
				printf("Il mazziere ha sballato.\nHai vinto con il punteggio di %.1f.", puntoGioc);
			}
		}else{
			printf("Hai perso!\nPartita vinta dal mazziere con il punteggio di %.1f.", puntoMazz);
		}
	}
	
return 0;
}

int matta (int cartaMatta, int flagMattaGioc){
	flagMattaGioc = 1;
	while(cartaMatta < 1 || cartaMatta > 7){
		puts("Quanto vuoi far valere la matta (da 1 a 7) ? ");
		scanf("%d", &cartaMatta);	
		if(cartaMatta < 1 || cartaMatta > 7){
			puts("Numero non valido! Inserire valore da 1 a 7 per continuare.");
		}
	}
	return cartaMatta;
}

float lancio (){
	float array[9] = {1,2,3,4,5,6,7,0.5,11};
	int indiceRandom = rand() % 9;
	return array[indiceRandom];
}

File: Programmi Java/SetteEMezzo.java
import java.util.*;

public class SetteEMezzo{
	public static void main(String args[]){
		double carte[] = {0.5, 1, 2, 3, 4, 5, 6, 7};
		boolean isMatta =  false; 
		boolean sballato = false;

		double carteRicevute = 0.0;
		double nuovaCarta = 0.0;
		double sommaCarte = 0.0;

		double carteCPU = 0.0;
		String risposta;

		Scanner input = new Scanner(System.in);
		Random rand = new Random();

		System.out.println("SETTE E MEZZO\n\n\n");

		//GIOCATA DELL'UTENTE
		carteRicevute = carte[rand.nextInt(8)];
		System.out.println("Hai ricevuto la carta: "+carteRicevute);

		//gestione del 4
		if(carteRicevute == 4){
			System.out.println("Vuoi cambiare carta?(si/no)");
			risposta = input.next();
			if(risposta.equals("si")){
				carteRicevute = carte[rand.nextInt(8)];
				System.out.println("\nHai ricevuto la carta: "+carteRicevute);
			}
		}
		
		//Gestione della matta
		if(carteRicevute == 0.5){
			isMatta = rand.nextBoolean();
			if(isMatta == true){
				System.out.println("Hai ricevuto la matta.");
				System.out.println("\nVuoi far valere la matta 0.5? (si/no)");
				risposta = input.next();
				if(risposta.equals("no")){
					do{
						System.out.println("\nQuanto vuoi farla valere? (1-7)");
						carteRicevute = input.nextDouble();
					}while(carteRicevute < 1 && carteRicevute > 7);
				}
			}
		}

		do{
			System.out.print("Vuoi un'altra carta? (si/no)\n-> ");
			risposta = input.next();

			if(risposta.equals("si")){
				nuovaCarta = carte[rand.nextInt(8)];
				if(isMatta == false){
					if(nuovaCarta == 0.5){
						isMatta = rand.nextBoolean();
						if(isMatta == true){
							System.out.println("E' uscita la matta.");
							System.out.println("\nVuoi far valere la matta 0.5? (si/no)");
							risposta = input.next();
							if(risposta.equals("no")){
								do{
									System.out.println("\nQuanto vuoi farla valere? (1-7)");
									nuovaCarta = input.nextDouble();
								}while(carteRicevute < 1 && carteRicevute > 7);
							}
						}
					}
				}

				carteRicevute += nuovaCarta;
				System.out.println("La somma delle carte attuali e': "+carteRicevute+"\n");
			}
		}while(carteRicevute < 7.5 && risposta.equals("si"));
	
		if(carteRicevute > 7.5){
			System.out.println("Hai sballato, vince il mazziere.");
			sballato = true;
		}else if(carteRicevute == 7.5){
			System.out.println("Hai fatto sette e mezzo, vediamo cosa fa il mazziere...");
		}

		//GIOCATA DELLA CPU
		if(sballato == false){
			do{
				carteCPU += carte[rand.nextInt(8)];
				System.out.println("Il mazziere fa: "+carteCPU);
			}while(carteCPU < carteRicevute);

			if(carteCPU > 7.5){
				System.out.println("\nIl mazziere ha sballato.");
				carteCPU = 0.0;
			}

			if(carteCPU > carteRicevute && carteCPU <= 7.5){
				System.out.println("\nMi dispiace hai perso!");
			}else if(carteCPU == carteRicevute){
				if(isMatta == true){
					System.out.println("\nComplimenti hai vinto!");
				}else{
					System.out.println("\nMi dispiace hai perso!");
				}
			}else if(carteRicevute > carteCPU && sballato == false){
				System.out.println("\nComplimenti hai vinto!");
			}
		}
	}
		
}